solana/modules/ntt-messages/src/trimmed_amount.rs

```diff
@@ -40,47 +42,81 @@ impl TrimmedAmount {
         Self { amount, decimals }
     }
 
-    pub fn change_decimals(&self, new_decimals: u8) -> Self {
+    pub fn change_decimals(&self, new_decimals: u8) -> Result<Self, ScalingError> {
         if new_decimals == self.decimals {
-            return *self;
+            return Ok(*self);
         }
-        Self {
-            amount: self.untrim(new_decimals),
+        let amount = self.untrim(new_decimals)?;
+        Ok(Self {
+            amount,
             decimals: new_decimals,
-        }
+        })
     }
 
-    fn scale(amount: u64, from_decimals: u8, to_decimals: u8) -> u64 {
+    fn scale(amount: u64, from_decimals: u8, to_decimals: u8) -> Result<u64, ScalingError> {
         if from_decimals == to_decimals {
-            return amount;
+            return Ok(amount);
         }
         if from_decimals > to_decimals {
-            amount / 10u64.pow((from_decimals - to_decimals).into())
+            // [`u64::checked_pow`] expects a u32 argument
+            let power: u32 = (from_decimals - to_decimals).into();
+            // Exponentiation will overflow u64 when `power` is greater than 18
+            let scaling_factor: u64 = match 10u64.checked_pow(power) {
+                Some(scaling_factor) => scaling_factor,
+                None => return Err(ScalingError::OverflowExponent),
```

Comment:
```suggestion
            let scaling_factor: u64 = 10u64.checked_pow(power).map_err(|_| ScalingError::OverflowExponent))?;
```

Reply:
I don't think I can use `map_err` here because `checked_pow` returns Option, not Result.

I'm going to try `ok_or` instead:
```rust
let scaling_factor: u64 = 10u64.checked_pow(power).ok_or(ScalingError::OverflowExponent)?;
```

Reply:
nice, yeah `ok_or` should work! my bad